mono/sgen/sgen-cardtable.c

```diff
@@ -441,18 +441,26 @@ sgen_card_table_clear_cards (void)
 }
 
 static void
-sgen_card_table_start_scan_remsets (void)
+sgen_card_table_start_scan_remsets (gboolean is_parallel, SgenObjectOperations *object_ops_nopar, SgenObjectOperations *object_ops_par)
 {
 #ifdef SGEN_HAVE_OVERLAPPING_CARDS
 	/*FIXME we should have a bit on each block/los object telling if the object have marked cards.*/
 	/*First we copy*/
+#ifdef SGEN_USE_SIMPLE_PAR_COPY_CLEAR_CARDS
```

Comment:
I think having this as a define is detrimental to giving the optimization more exposure. Normally we would use defines if the feature can have negative perf effects sometimes, even if not used, which doesn't seem like it is the case here. If you are uncomfortable having this by default in the parallel minor we can have another MONO_GC_PARAMS option, and have it enabled by default or through an environment variable on the console. Note that using this feature doesn't make sense if the GC is not parallel since we wouldn't have multiple sgen threadpool workers, so we should take that into account when the param is passed.

Reply:
I have seen mixed results depending on platform/hardware as well as size of major/los heaps, so maybe this should even be using some runtime heuristics to switch between options when running simple-par? Like checking against some size limits + time operation and switch if it takes more than some threshold value. It will complicate things for sure, but will then be more self adjusting and we could start with simple rule(s). The call will be coming in from collect_nursery that already have GC lock and running under STW, so I believe we could keep some state at that level, deciding if doing parallel start_scan_remsets or not when calling this method. Could that be an alternative?

Reply:
and we already have a timer on start_scan_remsets call in collect_nursery, so initially we could have just a bool not using parallel by default and if we hit some value, like 2-3 ms for this call, we could switch over parallelizing it when running with simple-par.

Reply:
For simplicity I'd leave heuristics out of this PR for now. Having it toggle-able at run time, based on gc param option, is good enough for now. Overall I have mixed feelings about toggling based on time since some cpu's are faster others are slower and the choice would seem very random. It would make more sense if we would toggle based on size. We could assume that a job should have a minimal set of memory that it needs to traverse, otherwise the overhead of enqueueing jobs is not worth it, for example. This heuristic would depend way less on the platform and feels cleaner.

Reply:
OK, size would be a component of heuristics as well, and yes piece is amount of memory to inspect. I convert it to a GC param option for now in this PR and maybe I experiment with heuristics on console platforms where we have known hardware and execution times.